### PR TITLE
Rework Interval type

### PR DESCRIPTION
**Pull request by robtimp**
I’d like to propose reworking the Interval type for a couple reasons:

* Creating a separate enum case for every possible interval does not scale well (intervals past a 15th are not currently included)
* Having separate cases will make it impractical to perform operations on intervals in the future (e.g. adding two intervals together)
* Doubly diminished and doubly augmented intervals, while not especially common, are a thing

### CODE REVIEW

**Comment by robtimp** [comment]
@drumnkyle @migue48 

Thoughts? Would you prefer an `init` that throws to a failable initializer?

File: MusicNotationKit/MusicNotationKit/Interval.swift [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    public var debugDescription: String {
+        var description = "\(quality.rawValue) "
+        switch number {
+        case 1:
+            description += "unison"
+        case 8:
+            description += "octave"
+        default:
+            let formatter = NumberFormatter()
+            if #available(OSX 10.11, *) {
```
Comment (drumnkyle) [inline MusicNotationKit/MusicNotationKit/Interval.swift]: I like that you used the if available, however I have avoided having any dependence on anything aside from Swift's standard library. I have not even relied on anything in Foundation. I'd like to keep it that way.

**Comment by drumnkyle** [comment]
Totally agree that the manual addition of intervals was crappy. I started doing it so I just finished it, but I wasn't sure if we were even going to use it. I definitely think this looks much better. I'm still not sure if we will be utilizing this information, but I guess it's good to have? What do you think @migue48

**Comment by drumnkyle** [comment]
And yes, I would prefer an init that throws. Thanks for checking.

**Comment by migue48** [comment]
I agree with this change. I was thinking to use this in a chord detection module.  Once we start implementing the [duration calculator changes](https://github.com/drumnkyle/music-notation-swift/blob/master/docs/duration-calculator.md) we may want to consider a similar approach for [NoteDuration](https://github.com/drumnkyle/music-notation-swift/blob/master/MusicNotationKit/MusicNotationKit/Types.swift?#L38).

**Comment by robtimp** [comment]
@drumnkyle @migue48 The Interval type would also be useful for transposition or inversion.

**Comment by drumnkyle** [comment]
Sounds good. Looks like we are all in agreement. @robtimp: can you make the changes from my comments and then I will merge it to master.

File: MusicNotationKit/MusicNotationKitTests/IntervalTests.swift [hunk 2]
```diff
@@ -12,76 +12,95 @@ import MusicNotationKit
...
+        let interval = try! Interval(quality: .perfect, number: 33)
         XCTAssertEqual(interval.debugDescription, "perfect 33rd")
         XCTAssertEqual(interval.abbreviation, "P33")
     }
     
     func testMajorOctaveInvalid() {
-        let interval = Interval(quality: .major, number: 8)
-        XCTAssertNil(interval)
+        do {
+            _ = try Interval(quality: .major, number: 8)
```
Comment (drumnkyle) [inline MusicNotationKit/MusicNotationKitTests/IntervalTests.swift]: Huh. I didn't know you didn't need a `let _` there.

File: MusicNotationKit/MusicNotationKitTests/IntervalTests.swift [hunk 3]
```diff
@@ -12,76 +12,95 @@ import MusicNotationKit
...
-        let interval = Interval(quality: .perfect, number: 9)
-        XCTAssertNil(interval)
-    }
+        do {
+            _ = try Interval(quality: .perfect, number: 9)
+            shouldFail()
+        } catch IntervalError.invalidQuality {
+        } catch {
+            expected(IntervalError.invalidQuality, actual: error)
+        }    }
```
Comment (drumnkyle) [inline MusicNotationKit/MusicNotationKitTests/IntervalTests.swift]: The last `}` should be on a new line.

**Comment by drumnkyle** [comment]
Thanks for making those changes. One more tiny fix before I merge it, then we are good.

**Pull request merged**